validate.go

```diff
@@ -0,0 +1,183 @@
+			container.Resources.Requests["memory"] = &newRequest
+			mutated = true
+		}
+	}
+
+	cpuStr, ok := container.Resources.Requests["cpu"]
+	if !ok || cpuStr == nil || len(strings.TrimSpace(string(*cpuStr))) == 0 {
+		if !settings.Cpu.DefaultRequestQuantity.IsZero() {
+			newRequest := api_resource.Quantity(settings.Cpu.DefaultRequestQuantity.String())
+			container.Resources.Requests["cpu"] = &newRequest
+			mutated = true
+		}
+	}
+	return mutated, nil
+}
+
+func validateAndAdjustContainerResourceLimits(container *corev1.Container, settings *Settings) (bool, error) {
+	mutated := false
+	memoryStr, ok := container.Resources.Limits["memory"]
+	if !ok || memoryStr == nil || len(strings.TrimSpace(string(*memoryStr))) == 0 {
+		if !settings.Memory.DefaultLimitQuantity.IsZero() {
+			newLimit := api_resource.Quantity(settings.Memory.DefaultLimitQuantity.String())
+			container.Resources.Limits["memory"] = &newLimit
+			mutated = true
+		}
+	} else {
+		memoryLimit, err := resource.ParseQuantity(string(*memoryStr))
+		if err != nil {
+			return false, fmt.Errorf("invalid memory limit")
+		}
+		if memoryLimit.Cmp(settings.Memory.MaxLimitQuantity) > 0 {
+			return false, fmt.Errorf("memory limit '%s' exceeds the max allowed value '%s'", memoryLimit.String(), settings.Memory.MaxLimitQuantity.String())
+		}
+	}
+
+	cpuStr, ok := container.Resources.Limits["cpu"]
+	if !ok || cpuStr == nil || len(strings.TrimSpace(string(*cpuStr))) == 0 {
+		if !settings.Cpu.DefaultLimitQuantity.IsZero() {
+			newLimit := api_resource.Quantity(settings.Cpu.DefaultLimitQuantity.String())
+			container.Resources.Limits["cpu"] = &newLimit
+			mutated = true
+		}
+	} else {
+		cpuLimit, err := resource.ParseQuantity(string(*cpuStr))
+		if err != nil {
+			return false, fmt.Errorf("invalid cpu limit")
+		}
+		if cpuLimit.Cmp(settings.Cpu.MaxLimitQuantity) > 0 {
+			return false, fmt.Errorf("cpu limit '%s' exceeds the max allowed value '%s'", cpuLimit.String(), settings.Cpu.MaxLimitQuantity.String())
+		}
+	}
+	return mutated, nil
+}
+
+func validateAndAdjustContainer(container *corev1.Container, settings *Settings) (bool, error) {
+	if container.Resources == nil {
+		container.Resources = &corev1.ResourceRequirements{
+			Limits:   make(map[string]*api_resource.Quantity),
+			Requests: make(map[string]*api_resource.Quantity),
+		}
+	}
+	if container.Resources.Limits == nil {
+		container.Resources.Limits = make(map[string]*api_resource.Quantity)
+	}
+	if container.Resources.Requests == nil {
+		container.Resources.Requests = make(map[string]*api_resource.Quantity)
+	}
+	limitsMutation, err := validateAndAdjustContainerResourceLimits(container, settings)
+	if err != nil {
+		return false, err
+	}
+	requestsMutation, err := validateAndAdjustContainerResourceRequests(container, settings)
+	if err != nil {
+		return false, err
+	}
+	return limitsMutation || requestsMutation, nil
+
+}
+func getImageAndTag(imageUri string) (string, string, error) {
```

Comment:
Parsing a docker reference is really complex, there are many corner cases.

I would use the same logic used by [the Gatekeeper policy](https://artifacthub.io/packages/gatekeeper/gatekeeper-policies/k8scontainerlimits) we're improving:

> Any container that uses an image that matches an entry in this list will be excluded
> from enforcement. Prefix-matching can be signified with `*`. For example: `my-image-*`.
> It is recommended that users use the fully-qualified Docker image name (e.g. start with a domain name)
> in order to avoid unexpectedly exempting images from an untrusted repository.

Please update the README to include this description.

The check can be done in a super simple way, like Rego does:

- If `settings.image_name` does not end with `*`: do a `==` comparison
- Else, remove the `*` from the `settings.image_name` and  check if `container.image_name.starts_with(image_without_star)`

Reply:
@flavio  I've implemented the change. As far as I can understand, we want to force the user to set the tag `*` if he/she wants to validate all the tags. Is that right?

In other words, if the workload has the image image:v1 and the settings has the image in the policy ignore list, the policy will validate the resources.

Reply:
No, it's not about setting `*` inside of the tag. This is about doing a prefix matching when the `*` is used.

This is the relevant part of the Rego policy:

```rego
is_exempt(container) {

    exempt_images := object.get(object.get(input, "parameters", {}), "exemptImages", [])

    img := container.image

    exemption := exempt_images[_]

    _matches_exemption(img, exemption)

}


_matches_exemption(img, exemption) {

    not endswith(exemption, "*")

    exemption == img

}


_matches_exemption(img, exemption) {

    endswith(exemption, "*")

    prefix := trim_suffix(exemption, "*")

    startswith(img, prefix)

}
```